Add time_per_element_counter to python SDK

boyuanzz opened the pull request:
Add time_counter behind time_per_element_counter experimental flag

What does time_counter do:
Collect processing_time per element for each step.

r: @pabloem @robertwb

boyuanzz commented:
I have several benchmark tests here and will add more benchmarks if possible:

1. Wordcount Pipeline(1gb inputs 10 times with 10 workers) with counter enabled:
```
+------------------+
| PipelineDuration |
+------------------+
| 691              |
| 722              |
| 692              |
| 746              |
| 767              |
| 742              |
| 721              |
| 736              |
| 736              |
| 701              |
+------------------+
AVG(feature branch runtime) = 725.4s
AVG(master branch) = 681s
```
2. Benchmarks got from @robertwb 's script( ran 10 times each):
```
With counter enable:
AVG(per_element_cost) = 8.92266249657e-07
```
```
With counter disabled:
AVG(per_element_cost) = 8.34413099289e-07
```
```
origin:master branch:
AVG(per_element_cost)  = 8.63162302971e-07
```

boyuanzz commented:
jenkins retest this please

aaltay commented:
Two high level comments:
1. You asked 4 reviewers to review, do you need a review from all of them? I will wait for first reviewer to make a few passes first.
2. 6% regression on wordcount benchmark is high. Could you also look at large benchmarks to see the impact.

boyuanzz commented:
Hey @aaltay,
Thanks so much for your comment! 
For comment1:
Robert and Charles are required reviewers, also Pablo is required from our team.
For comment2:
After distribution counter optimization merged(https://github.com/apache/beam/pull/4928), time_counter will run into 700ms averagely  for the same wordcount pipeline, which is less than 3% regression. I'll also try to run other benchmark tests.

pabloem commented:
I'm taking a look now, so others don't need to look yet.

pabloem commented on sdks/python/apache_beam/runners/worker/statesampler_fast.pyx:
```diff
@@ -101,6 +103,13 @@ cdef class StateSampler(object):
     pythread.PyThread_release_lock(self.lock)
+    self.element_tracker = DataflowElementExecutionTracker()
+    # TODO: remove experimental flag once time_counter release
+    experiments = RuntimeValueProvider.get_value('experiments', str, '')
```
You can use experiments [like this](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/worker/operations.py#L292). (I fixed my earlier bug :))

boyuanzz replied:
Changed in latest commit. Thanks for fix!

pabloem commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,192 @@
+#
+
+# cython: profile=True
+
```
Please document the classes in this file in detail, including concurrency / locking expectations, and a sketch for the algorithm if possible.
I'll continue to review these in the meantime.

boyuanzz replied:
I'll add them in next revision. For sketch of the algorithm, maybe it's better to demo in unit tests. I have a typical usage test case in corresponding unit test and I'll add more comments in that unit test.

pabloem commented:
Also please fix py27, py27gcp, lint and other tests.

pabloem commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,192 @@
+
+  cdef void add_journal(self, ElementExecutionPtr execution_ptr,
+                        int64_t snapshot):
+    if snapshot <= self.max_snapshot:
```
<!--new_thread; commit:1f7d05e01de972b0e3bec788fe9c82377ea44f15; resolved:0-->
It seems that it should be called min_snapshot? : )

boyuanzz replied:
Actually it's max_snapshot since journal keeps track of the increasing snapshot number and always keep the largest one.

pabloem commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker_test.py:
```diff
@@ -0,0 +1,200 @@
+        import DataflowElementExecutionTracker
+      self.tracker = DataflowElementExecutionTracker()
+    except ImportError:
+      self.tracker = None
```
<!--new_thread; commit:1f7d05e01de972b0e3bec788fe9c82377ea44f15; resolved:0-->
You can use SkipTest to skip tests that are not in Cython (https://github.com/apache/beam/blob/b095837ff7ee05a9f5b1282aae783ceccce7ec99/sdks/python/apache_beam/runners/worker/statesampler_test.py#L38)

boyuanzz replied:
Fixed this in the latest commit, thanks!

boyuanzz commented:
Hey @charlesccychen, could you please take a look at this code review? Pablo has given LGTM in first iteration. Thanks!

charlesccychen commented:
Thanks Boyuan for this change.  I read through the changes here, but before we get into the details, do you have a good sense of what is causing the ~6% wordcount performance regression (if the numbers are accurate, the actual processing time overhead may be larger due to Dataflow worker fixed costs)?  My first guess would be the per-element cost entering / exiting the scoped state, but the unchanged mapper microbenchmark numbers you listed seem to suggest that is not the case.  Do you think, then, that this performance regression is coming from the main sampling loop?  We should figure out what the bottleneck is so that we can focus on the most impactful way to improve it.

boyuanzz commented:
Commit https://github.com/apache/beam/pull/4925/commits/ccf5366d43c5a977dabf49a84a65d338f017162e: Migrate new DataflowDistributionCounter(https://github.com/apache/beam/commit/c4d2d7d717767b803864d788396e4d3dbd419cff) into time_counter. @aaltay Please review this new migration.

Commit https://github.com/apache/beam/pull/4925/commits/bdfbf7813f29cfda39312640547d26ca30bbc5eb: Make run_tox_cleanup.sh clean *.cpp ext also.

charlesccychen commented:
@boyuanzz: Can you please provide benchmarks after the rebase to see if the regression is still present?

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+        break
+    # Keep the currently executing element in the journal.
+    # So its remaining execution time is counted in the next sampling round.
+    if snapshotted_ptr != NULL and snapshotted_ptr.snapshot == latest_snapshot:
```
Ah...I'm afraid not. For last execution item, we want to attribute processing time to this execution and we also want to keep this execution in the journal until next sampling. The motivation to do the thing like this is, even though we are in a sampling  manner, we still want to make attributing time as precious as possible.

aaltay replied:
This is fine either way. For simplicity's sake, what do we lose if we filter by <, would not the same item would show up in metrics in the next round anyway?

boyuanzz replied:
Compared to ```<=```, ```<``` may cause some execution get less attribution time.

aaltay replied:
> may cause some execution get less attribution time.

Is this temporary? And would it fix itself in the next round of counter updates?

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+    cdef int64_t attribution_time = 0
+    # Attribute processing time.
+    while it != executions_per_element.end():
+      element_ptr = dref(it).first
```
I'll review all comments to make them clearer. Thanks for pointing out!

aaltay replied:
Thank you.

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+    """Create and journal a new ElementExecution to track a processing element.
+    """
+    cdef ElementExecutionPtr execution_ptr = (
+      <ElementExecutionPtr> malloc(sizeof(ElementExecution)))
```
I'm not sure what a ```pool``` means here. The reasons why I alloc memory dynamically are: 
1. It's faster to move pointer here and there rather than copying item directly.
2. At this point, we may need memory alloced in heap rather than stack.

aaltay replied:
I was thinking that we can allocate a larger object in the heap first, and use indexes into that rather than doing bunch of small allocations.

> It's faster to move pointer here and there rather than copying item directly.

Note that, the item (`ElementExecution`) is `char* operation_name` is just a struct wrapping around a `char*`. We might as well pass pointers to the operation_name instead of allocating another object. Do we plan on adding more items into `ElementExecution`?

boyuanzz replied:
For ```ElementExecution```:
1. We don't have a plan adding more attr into this struct.
2. The reason why we need to keep a ```ElementExecution``` is, we want to track processing time per element. So the ```ElementExecutionPtr``` actually is a object that marks an element with operaion_name.
3. If we can do the pre-allocating things, that would be better. I'm concerning, we cannot know the size of pool at very beginning, which means we may need resize the pool at some point.

aaltay replied:
I think you can still use `ElementExecutionPtr` by typedefing it to `char*`, would it work?

Let's not do 3 in this PR. One can do malloc's to extend the size of this large object when needed. This will happen much less frequently than individually allocation memory for each element.

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+    self.execution_stack.pop_back()
+    cdef int64_t next_snapshot = self.shared_state.latest_snapshot + 1
+    self.shared_state.done_journal.add_journal(last_execution, next_snapshot)
+    if self.execution_stack.empty() is False:
```
Oh right, we don't need this if block here.

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pxd:
```diff
@@ -0,0 +1,89 @@
+# limitations under the License.
+#
+
+# cython: profile=False
```
I guess time_counter will be more time-consuming (especially, take_sample function call will call every time statesampler wakes up). This is the major reason I wanna disable profiling here. How do you feel like?

aaltay replied:
I think we need to make a global decision and not a file by file decision in a PR. It is fair to bring this up as dev@ discussion and we can discuss it. 

For this PR, does how much does it add to regression to leave profiling on?

charlesccychen replied:
I asked @robertwb about the `cython: profile` directive in the past, and he indicated that the overhead here was very small, on the order of a dozen machine instructions.  We should keep it on.

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+      raise ValueError('Timestamps must be monotonically increasing.'
+                       + 'Input snapshot %d is not greater than max snapshot %d'
+                       % (snapshot, self.max_snapshot))
+    cdef SnapshottedExecutionPtr snapshotted_ptr = (
```
I came up with 2 solutions here:
1. Once malloc return NULL, skip all following possible steps.
2. Have something like retry to alloc a fixed number of times. If fail to alloc anyway, then skip following steps.
Which one you think would be better here?

aaltay replied:
I think it is better to not retry. How about asserting and failing in a predictable way. I do not know what would be the meaning of skipping following steps. In that case we need to communicate to the caller that `add_journal` failed, and caller needs to do something about it.

boyuanzz replied:
```Assert``` would be better. I'll update in next commit. Thanks~!

boyuanzz commented:
Commit https://github.com/apache/beam/pull/4925/commits/f08d3963c2b6c12f9bbc3acfa7ef9fb679178d04: Fixed possible memory leak cased by dynamic memory allocation.

Commit  https://github.com/apache/beam/pull/4925/commits/d3696d30769f26d962a330a27d8ad0e14a474244 and https://github.com/apache/beam/pull/4925/commits/0e6bd75340c962c025d50bdba71afb3a77c614ba:
Addressed @aaltay 's comments, but there are some unaddressed comments which need more discussions.

aaltay commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+    until the next sampling round in order to account timing for the final 
+    fragment of execution.
+    """
+    if self.execution_stack.size() <= 1:
```
So you are saying both cases are an error in an equal way. Sounds good in that case.

boyuanzz commented:
Commit https://github.com/apache/beam/pull/4925/commits/0497ea53e65a019766f30e4031e42726cd75903f:
Add ```assert``` statement in case ```malloc``` failed. @aaltay

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+        break
+    # Keep the currently executing element in the journal.
+    # So its remaining execution time is counted in the next sampling round.
+    if snapshotted_ptr != NULL and snapshotted_ptr.snapshot == latest_snapshot:
```
<!--thread_id:cc_182316149_t; commit:bdfbf7813f29cfda39312640547d26ca30bbc5eb; resolved:0-->
<!--section:context-quote-->
> **aaltay** wrote:
> > may cause some execution get less attribution time.
> 
> Is this temporary? And would it fix itself in the next round of counter updates?

<!--section:body-->
Unfortunately no. ```<``` will cause some execution get less attributing time in final result.

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+    """Create and journal a new ElementExecution to track a processing element.
+    """
+    cdef ElementExecutionPtr execution_ptr = (
+      <ElementExecutionPtr> malloc(sizeof(ElementExecution)))
```
<!--thread_id:cc_182317472_t; commit:bdfbf7813f29cfda39312640547d26ca30bbc5eb; resolved:0-->
<!--section:context-quote-->
> **aaltay** wrote:
> I think you can still use `ElementExecutionPtr` by typedefing it to `char*`, would it work?
> 
> Let's not do 3 in this PR. One can do malloc's to extend the size of this large object when needed. This will happen much less frequently than individually allocation memory for each element.

<!--section:body-->
I'm afraid  ```typedef ElementExecutionPtr char*``` will not work, unless every time we ```malloc``` a new ```char*``` and do ```memcpy``` op_name to this new address.

aaltay replied:
I am a bit confused here.

Currently: ElementExecution is a struct that has a single char*. We do not take ownership of the char*, or neither we need to memcpy contents of this char*. We malloc a new ElementExecution.

If instead we use typedef ElementExecutionPtr, we can exactly do the same thing. We still do not need to take ownership of char* or memcpy it.

Could you explain why do we need to do the copying in the latter case?

boyuanzz replied:
Are you asking, why we need to create a new object to wrap char* here? If so, we use a new object to mark one element(based on different address, we use this address to group the same element later on), since it's possible that there are many elements going through one operation. We want to calculate processing time per element per operation.

aaltay replied:
Maybe we can have video chat to settle this out.

Yes, my question is why do we need to create a new object. It seems like the char* itself is enough to uniquely identify the element.

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pxd:
```diff
@@ -0,0 +1,89 @@
+# limitations under the License.
+#
+
+# cython: profile=False
```
<!--thread_id:cc_182313240_t; commit:bdfbf7813f29cfda39312640547d26ca30bbc5eb; resolved:0-->
<!--section:context-quote-->
> **boyuanzz** wrote:
> I'll turn it on in next commit. Thanks for providing info.

<!--section:body-->
https://github.com/apache/beam/pull/4925/commits/6597b824664b808f00fb93b44a55097d9cdd38ad turns profiling on.

boyuanzz commented on sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx:
```diff
@@ -0,0 +1,330 @@
+    """Create and journal a new ElementExecution to track a processing element.
+    """
+    cdef ElementExecutionPtr execution_ptr = (
+      <ElementExecutionPtr> malloc(sizeof(ElementExecution)))
```
<!--thread_id:cc_182317472_t; commit:bdfbf7813f29cfda39312640547d26ca30bbc5eb; resolved:0-->
<!--section:context-quote-->
> **aaltay** wrote:
> Maybe we can have video chat to settle this out.
> 
> Yes, my question is why do we need to create a new object. It seems like the char* itself is enough to uniquely identify the element.

<!--section:body-->
Discussed offline.

aaltay commented:
@boyuanzz What is the status of this PR? What are the next actions?

boyuanzz commented:
This PR is ready to merge, but blocked by performance issue.

boyuanzz commented:
Hey @aaltay, please review the latest changes: https://github.com/apache/beam/pull/4925/commits/f7a2669ff7bd6eb2b74d5d1d31e65ca2fe798c5d , thanks!

aaltay commented on sdks/python/apache_beam/options/value_provider.py:
```diff
@@ -69,6 +69,7 @@ def __eq__(self, other):
 
 class RuntimeValueProvider(ValueProvider):
   runtime_options = None
+  is_time_counter_enabled = False
```
@boyuanzz why do you need this to be a special case here? `RuntimeValueProvider` is not the right place to have attributes for different options.

Does it make a difference in performance? Can you cache this somewhere else?

boyuanzz replied:
For this PR, there is no significant difference in performance. Could you please take a look at the latest 2 commits in this https://github.com/apache/beam/pull/4741 ? 
My idea here is:
Every time trying to check experimental options by calling [RuntimeValueProvider.get_value()](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/options/value_provider.py#L82), it will cost O(n), where n is the # of options. For https://github.com/apache/beam/pull/4741 it's kind of necessary since we call it per element. For this PR, I wanna keep the same pattern. 
Maybe it's the best idea to do this, could you explain more about why 'RuntimeValueProvider is not the right place to have attributes for different options.' ?

aaltay commented on sdks/python/apache_beam/options/value_provider.py:
```diff
@@ -101,6 +102,11 @@ def get(self):
   @classmethod
   def set_runtime_options(cls, pipeline_options):
     RuntimeValueProvider.runtime_options = pipeline_options
+    # Set is_time_counter_enabled = True if 'time_per_element_instrumentation'
```
You can drop the comments here, it is exactly describing what the code does and does offer additional context to the reader.

boyuanzz replied:
Deleted.

boyuanzz commented:
Revert RuntimeValueProvider changes. 
Next action:
Wait for https://github.com/apache/beam/pull/5308 merged, then do corresponding changes.

boyuanzz commented:
retest this please

boyuanzz commented:
retest this please

boyuanzz commented:
Hey @aaltay, please review the latest commit https://github.com/apache/beam/pull/4925/commits/bf11de3f7ff02b81053b978895bb819ed5c85bf2. This commit changed the way reading experimental flag into the latest pattern: https://github.com/apache/beam/commit/60f90c8dcb229c35a82c7be15e64a89678bae058

boyuanzz commented:
This PR is ready for merge. Please merge this one  after 2.5.0 release branch cut.

pabloem commented:
This PR is also meant for inclusion. Boyuan, sorry for the trouble, but could you rebase it please? : )

boyuanzz commented:
Will resolve conflict asap. Thanks @pabloem
